backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramCreationService.java

```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2020 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sirius.web.collaborative.diagrams.api;
+
+import java.util.Optional;
+
+import org.eclipse.sirius.web.diagrams.Diagram;
+import org.eclipse.sirius.web.diagrams.description.DiagramDescription;
+import org.eclipse.sirius.web.services.api.objects.IEditingContext;
+
+/**
+ * Service used to create diagrams from scratch or from an existing diagram.
+ *
+ * @author sbegaudeau
+ */
+public interface IDiagramCreationService {
+
+    /**
+     * Creates a new diagram using the given parameters.
+     *
+     * @param label
+     *            The label of the diagram
+     * @param targetObject
+     *            The object used as the target
+     * @param diagramDescription
+     *            The description of the diagram
+     * @param editingContext
+     *            The editing context
+     * @return A new diagram properly layouted and saved in the data store
+     */
+    Diagram create(String label, Object targetObject, DiagramDescription diagramDescription, IEditingContext editingContext);
+
+    /**
+     * Refresh an existing diagram.
+     *
+     * <p>
+     * Refreshing a diagram since to always be possible but it may not be the case. In some situation, the semantic
+     * element on which the previous diagram has been created may not exist anymore and thus we can return an empty
+     * optional if we are unable to create the new diagram.
```

Comment:
I am not a fan of the name IDiagramCreationService for something that either create from scratch or refresh a diagram but I have not found a better name yet (we could split it in two IDiagramCreationService and IDiagramRefreshService, I don't mind). I agree with you on the fact that we are mixing a logical diagram and a diagram state when we are talking about "diagrams".

I disagree with you on the fact that IDiagramContext should be merged with RepresentationDescriptor. RepresentationDescriptor is used to persist representations while IDiagramContext is only used to give a diagram state to a tool and retrieve a diagram state after the execution of said tool (it may be the same, it may has changed just like the editing context). The persistence is not involved at anytime during this process.

I also disagree on a very minor point, our current Diagram is a diagram state but I think that if we had to introduce a DiagramState concept it would carry additional information and thus wouldn't be a simple renaming of Diagram, something like that:

```
@Immutable
class DiagramState {
  createdBy: UUID
  createdOn: Date
  state: Diagram // The current diagram
}
```

On another side, RepresentationDescriptor is one of the worst concept we currently have in our project. It has been created in a hurry and has never been well specified. The fact that the label is part of both the root of representation description and part of the representation contained by representation descriptor is just the result of something decided in a hurry to ship a new feature. Label is part of a large amount of information which would be needed for presentation purpose but only this one has really been used for now so we have duplicated it. The whole class is a mess. Label should never have been duplicated on RepresentationDescriptor.

I completely agree with you on the fact that RepresentationDescriptor is the identity of the representation, stable over time, which contains a reference to the latest representation state. It could even contain a reference to all previous representation states to provide some version support. It is not good at doing this job right now. I don't see why we should give tools access to the RepresentationDescriptor.

Tools can create a new diagram state but while they can do anything they want, it may not be coherent. They could for example delete synchronized mappings. We will just take their created diagram state and use it as an input to refresh the real diagram state and produce a new diagram state safely that we will store in the representation descriptor.